Support emitting calli on function pointer invocations.

333fred opened the pull request:
@AlekseyTs @dotnet/roslyn-compiler for review.

333fred commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -433,6 +438,928 @@ unsafe struct S
+}");
+        }
+
+        [Theory(Skip = "PROTOTYPE(func-ptr)")]
```
I was having some issues figuring out the required il for this test and wanted to get the PR out before I left, so I skipped it for now.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/ITokenDeferral.cs:
```diff
@@ -12,13 +12,14 @@ internal interface ITokenDeferral
 
         string GetStringFromToken(uint token);
-        Cci.IReference GetReferenceFromToken(uint token);
+        object GetReferenceFromToken(uint token);
```
>GetReferenceFromToken [](start = 15, length = 21)

Consider adding a doc comment listing the types that can come out of this method.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/ReferenceDependencyWalker.cs:
```diff
@@ -145,6 +145,25 @@ private static void VisitMethodReference(Cci.IMethodReference methodReference, E
             }
         }
 
+        internal static void VisitSignature(Cci.ISignature signature, EmitContext context)
```
>VisitSignature [](start = 29, length = 14)

Consider adjusting the previous method to share the code with this one, by calling this method for the same set of operations.

333fred replied:
> Consider adjusting the previous method to share the code with this one, by calling this method for the same set of operations.

The reason I didn't do that is because that intersperses the ExtraArguments check, which `ISignature` does not have.

---
In reply to: [382332630](https://github.com/dotnet/roslyn/pull/41794#discussion_r382332630) [](ancestors = 382332630)

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/ReferenceDependencyWalker.cs:
```diff
@@ -145,6 +145,25 @@ private static void VisitMethodReference(Cci.IMethodReference methodReference, E
             }
         }
 
+        internal static void VisitSignature(Cci.ISignature signature, EmitContext context)
```
>VisitSignature [](start = 29, length = 14)

Do we have a test that makes the work done in this method observable? It would be good to have a NoPia test, for example.

AlekseyTs replied:
>VisitSignature [](start = 29, length = 14)

It feels like we should be calling this method for function pointer types in VisitTypeReference method, otherwise we won't record references in declarations.

AlekseyTs commented on src/Compilers/Core/Portable/CodeGen/TokenMap.cs:
```diff
@@ -20,20 +21,19 @@ namespace Microsoft.CodeAnalysis.CodeGen
     /// the method bodies will be patched.
     /// To support these two scenarios we need two maps - Item-->uint, and uint-->Item.  (the second is really just a list).
     /// </summary>
```
></summary> [](start = 8, length = 10)

Consider including information what types can be in the map

AlekseyTs commented on src/Compilers/CSharp/Portable/Emitter/Model/FunctionPointerTypeSymbolAdapter.cs:
```diff
@@ -31,5 +45,37 @@ internal sealed partial class FunctionPointerTypeSymbol : IFunctionPointerTypeRe
         IDefinition? IReference.AsDefinition(EmitContext context) => null;
+
+        /// <summary>
+        /// We need to be able to differentiate between a FunctionPointer used a type and function pointer used
```
>used a type [](start = 74, length = 11)

"used **as** a type"?

AlekseyTs commented:
>                 return sx.Equals(sy, _compareKind);

Do we need to provide a way to pay attention to ```_compareKind``` when we compare FunctionPointerMethodSignature instances here?

---
Refers to: src/Compilers/Core/Portable/Emit/MetadataEntityReferenceComparer.cs:33 in 5a4c5b9. [](commit_id = 5a4c5b954514c4b17886c88f6097b7b8e1a88626, deletion_comment = False)

AlekseyTs commented on src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs:
```diff
@@ -3025,28 +3037,32 @@ private static int ReadInt32(ImmutableArray<byte> buffer, int pos)
-                _referenceVisitor.VisitMethodBodyReference(reference);
+                if (entity is IReference reference)
+                {
+                    _referenceVisitor.VisitMethodBodyReference(reference);
```
>_referenceVisitor.VisitMethodBodyReference(reference); [](start = 20, length = 54)

It feels like we want to do the same for ISignature here.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -329,7 +329,16 @@
+    <!-- Non-null type is required for this node kind -->
+    <Field Name="Type" Type="TypeSymbol" Override="true" Null="disallow"/>
+    <Field Name="InvokedExpression" Type="BoundExpression"/>
+    <Field Name="FunctionPointer" Type="FunctionPointerTypeSymbol"/>
```
><Field Name="FunctionPointer" Type="FunctionPointerTypeSymbol"/> [](start = 4, length = 64)

Wouldn't this be the type of ```InvokedExpression```?

333fred replied:
Yes, we could just use that.

---
In reply to: [382340027](https://github.com/dotnet/roslyn/pull/41794#discussion_r382340027) [](ancestors = 382340027)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:
```diff
@@ -860,17 +865,21 @@ private void ReportDuplicateNamedArgument(MemberResolutionResult<TMember> result
             NamedTypeSymbol typeContainingConstructor,
-            NamedTypeSymbol delegateTypeBeingInvoked)
+            NamedTypeSymbol delegateTypeBeingInvoked,
+            FunctionPointerMethodSymbol functionPointerMethodBeingInvoked)
```
>functionPointerMethodBeingInvoked [](start = 40, length = 33)

Instead of having consumer pass this argument, could we simply do a type check on what we have in ```symbols```?

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Semantics/AccessCheck.cs:
```diff
@@ -161,6 +161,23 @@ internal static bool IsEffectivelyPublicOrInternal(Symbol symbol, out bool isInt
+                    var funcPtr = (FunctionPointerTypeSymbol)symbol;
+                    if (!IsSymbolAccessibleCore(funcPtr.Signature.ReturnType, within, throughTypeOpt: null, out failedThroughTypeCheck, compilation, ref useSiteDiagnostics, basesBeingResolved))
+                    {
+                        return false;
```
>return false; [](start = 24, length = 13)

Do we report a diagnostics that says which specific type is not accessible?

333fred replied:
> Do we report a diagnostics that says which specific type is not accessible?

No. But as far as I can tell, there isn't actually a scenario today where this can return false. The only types of symbols allowed in the return type or parameter types are actual TypeSymbols, and we unconditionally return true for all of those types of symbols right below.

---
In reply to: [382344495](https://github.com/dotnet/roslyn/pull/41794#discussion_r382344495) [](ancestors = 382344495)

AlekseyTs replied:
> But as far as I can tell, there isn't actually a scenario today where this can return false. The only types of symbols allowed in the return type or parameter types are actual TypeSymbols, and we unconditionally return true for all of those types of symbols right below.

What about the ```case SymbolKind.NamedType:``` and alike above?

---
In reply to: [384858806](https://github.com/dotnet/roslyn/pull/41794#discussion_r384858806) [](ancestors = 384858806,382344495)

333fred replied:
True, but we have the same behavior here as generic parameters of NamedTypeSymbols. I'll add a test demonstrating inaccessible error messages.

---
In reply to: [387967547](https://github.com/dotnet/roslyn/pull/41794#discussion_r387967547) [](ancestors = 387967547,384858806,382344495)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:
```diff
@@ -1643,5 +1651,84 @@ private bool InvocableNameofInScope()
+                    funcPtr,
+                    analyzedArguments.Arguments.SelectAsArray((expr, args) => args.binder.BindToNaturalType(expr, args.diagnostics), (binder: this, diagnostics)),
+                    analyzedArguments.RefKinds.ToImmutableOrNull(),
+                    CreateErrorType(),
```
>CreateErrorType() [](start = 20, length = 17)

```funcPtr.Signature.ReturnType```? Since there is only one candidate, we can use return type for error recovery.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/CodeGenerator.cs:
```diff
@@ -348,6 +348,11 @@ private void EmitSymbolToken(FieldSymbol symbol, SyntaxNode syntaxNode)
             _builder.EmitToken(_module.Translate(symbol, syntaxNode, _diagnostics), syntaxNode, _diagnostics);
         }
 
+        private void EmitSymbolToken(FunctionPointerTypeSymbol symbol, SyntaxNode syntaxNode)
```
>private void EmitSymbolToken(FunctionPointerTypeSymbol symbol, SyntaxNode syntaxNode) [](start = 8, length = 85)

It looks like if we call ```private void EmitSymbolToken(TypeSymbol symbol, SyntaxNode syntaxNode)``` with the same arguments, we get something different. Consider changing the name of this method to make it clear that it is not an overload.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:
```diff
@@ -3392,5 +3412,49 @@ private void EmitBox(TypeSymbol type, SyntaxNode syntaxNode)
+            FunctionPointerMethodSymbol method = ptrInvocation.FunctionPointer.Signature;
+            EmitArguments(ptrInvocation.Arguments, method.Parameters, ptrInvocation.ArgumentRefKindsOpt);
+            var stackBehavior = GetCallStackBehavior(ptrInvocation.FunctionPointer.Signature, ptrInvocation.Arguments);
+            EmitExpression(ptrInvocation.InvokedExpression, used: true);
```
>EmitExpression(ptrInvocation.InvokedExpression, used: true); [](start = 12, length = 60)

It looks like this is going to violate an evaluation order. Do we have a test asserting the evaluation order, i.e. the target is evaluated before the arguments?

333fred replied:
Yes, we have multiple. The LocalRewriter change is responsible for spilling the InvokedExpression if necessary to ensure evaluation happens in the correct order.

---
In reply to: [382381636](https://github.com/dotnet/roslyn/pull/41794#discussion_r382381636) [](ancestors = 382381636)

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_FunctionPointerInvocation.cs:
```diff
@@ -0,0 +1,54 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
```
>using [](start = 0, length = 5)

```#nullable enable``` ?

333fred replied:
Won't be relevant in next revision.

---
In reply to: [382382617](https://github.com/dotnet/roslyn/pull/41794#discussion_r382382617) [](ancestors = 382382617)

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_FunctionPointerInvocation.cs:
```diff
@@ -0,0 +1,54 @@
+                    var temp = _factory.StoreToTemp(visitedInvokedExpression, out BoundAssignmentOperator store);
+                    stores.Add(store);
+                    temps.Add(temp.LocalSymbol);
+                    visitedInvokedExpression = temp;
```
>visitedInvokedExpression = temp; [](start = 20, length = 32)

The node fully defines semantics, Emit layer should enforce the right order of evaluation instead of assuming that some other layer took care of that.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_FunctionPointerInvocation.cs:
```diff
@@ -0,0 +1,54 @@
+                case var _ when node.Arguments.IsEmpty:
+                    stores.Free();
+                    temps.Free();
+                    return updateNode(node, visitedInvokedExpression, visitedArguments);
```
>return updateNode(node, visitedInvokedExpression, visitedArguments); [](start = 20, length = 68)

What if arguments mutate the local or parameter?

AlekseyTs commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:
```diff
@@ -264,6 +264,26 @@ public override void VisitMethod(IMethodSymbol symbol)
+
+                foreach (var param in symbol.Parameters)
+                {
+                    param.Accept(this);
```
>this [](start = 33, length = 4)

```this.NotFirstVisitor```?

AlekseyTs commented on src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Members.cs:
```diff
@@ -264,6 +264,26 @@ public override void VisitMethod(IMethodSymbol symbol)
+                    AddPunctuation(SyntaxKind.CommaToken);
+                }
+
+                symbol.ReturnType.Accept(this);
```
>this [](start = 41, length = 4)

```this.NotFirstVisitor```?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TypeSymbolExtensions.cs:
```diff
@@ -335,6 +335,12 @@ public static bool IsSZArray(this TypeSymbol type)
 
+        public static bool IsFunctionPointerType(this TypeSymbol type)
+        {
+            RoslynDebug.Assert((object)type != null);
```
>RoslynDebug.Assert((object)type != null); [](start = 12, length = 41)

It is not clear what is the purpose of the assert.

333fred replied:
I'm following the same conventions as the rest of the `IsX` above.

---
In reply to: [382385354](https://github.com/dotnet/roslyn/pull/41794#discussion_r382385354) [](ancestors = 382385354)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -433,6 +438,928 @@ unsafe struct S
+{
+    static void Main()
+    {
+        var retValue = Program.LoadPtr()(""Unused"");
```
>var retValue =  [](start = 8, length = 15)

Please remove variable declaration so that the value was unambiguously unused.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -433,6 +438,928 @@ unsafe struct S
+    {
+        Program.field = ""Field"";
+        delegate*<ref string> pointer = Program.LoadPtr();
+        Console.WriteLine(pointer());
```
>pointer() [](start = 26, length = 9)

How is this unused? Consider renaming the test to RefReturnUseByVal or something similar.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -433,6 +438,928 @@ unsafe struct S
+}");
+        }
+
+        // typeof() on function pointer
```
>// typeof() on function pointer [](start = 8, length = 31)

Is this comment about the test added above?

AlekseyTs commented:
Done with review pass (iteration 2)

333fred commented:
>                 return sx.Equals(sy, _compareKind);

I'll add a prototype comment to IFunctionPointerMethodSignature. I think in order to address this we'll have to factor equality out into a separate internal interface as I don't think we should be implementing ISymbolInternal on IFunctionPointerMethodSignature, so I'll address that in a followup when nullability is addressed in general.

---
In reply to: [589440631](https://github.com/dotnet/roslyn/pull/41794#issuecomment-589440631) [](ancestors = 589440631)

---
Refers to: src/Compilers/Core/Portable/Emit/MetadataEntityReferenceComparer.cs:33 in 5a4c5b9. [](commit_id = 5a4c5b954514c4b17886c88f6097b7b8e1a88626, deletion_comment = False)

333fred commented:
@AlekseyTs addressed feedback.

---
In reply to: [589769629](https://github.com/dotnet/roslyn/pull/41794#issuecomment-589769629) [](ancestors = 589769629)

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/CodeGenerator.cs:
```diff
@@ -348,6 +348,11 @@ private void EmitSymbolToken(FieldSymbol symbol, SyntaxNode syntaxNode)
             _builder.EmitToken(_module.Translate(symbol, syntaxNode, _diagnostics), syntaxNode, _diagnostics);
         }
 
+        private void EmitFunctionPointerSymbolToken(FunctionPointerTypeSymbol symbol, SyntaxNode syntaxNode)
```
>EmitFunctionPointerSymbolToken [](start = 21, length = 30)

Consider using "EmitSignatureToken" as the name. For me "FunctionPointer" still sounds like this is about a type, i.e. about a type token.

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/CodeGenerator.cs:
```diff
@@ -348,6 +348,11 @@ private void EmitSymbolToken(FieldSymbol symbol, SyntaxNode syntaxNode)
 
+        private void EmitFunctionPointerSymbolToken(FunctionPointerTypeSymbol symbol, SyntaxNode syntaxNode)
+        {
+            _builder.EmitToken(((Cci.IFunctionPointerTypeReference)symbol).Signature, syntaxNode, _diagnostics);
```
>symbol [](start = 67, length = 6)

It feels like we should call translate as we do in other similar helpers

AlekseyTs commented on src/Compilers/CSharp/Portable/CodeGen/Optimizer.cs:
```diff
@@ -1671,6 +1675,23 @@ public override BoundNode VisitReturnStatement(BoundReturnStatement node)
+            {
+                // This local's side effects must come before the argument's side effects, but it must
+                // be on top of the stack when calling the function pointer
+                ShouldNotSchedule(localSymbol);
```
>ShouldNotSchedule(localSymbol); [](start = 16, length = 31)

Do we still need this special handling?

AlekseyTs commented:
Done with review pass (iteration 3)

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenFunctionPointersTests.cs:
```diff
@@ -430,7 +441,1842 @@ unsafe struct S
+            var aRef = CreateCompilation(@"
+using System.Runtime.CompilerServices;
+[assembly: InternalsVisibleTo(""B"")]
+internal class A {}", assemblyName: "A").EmitToImageReference();
```
>internal [](start = 0, length = 8)

Does the error go away if the type made public? Could you please add a test for this scenario as well?

333fred commented:
@dotnet/roslyn-compiler for a second review please

tannergooding commented:
I tested this out locally on my DirectX 12 bindings and everything works 🎉:
![image](https://user-images.githubusercontent.com/10487869/75997946-f40f7780-5eb4-11ea-8d66-d9260bcfcd67.png)

333fred commented on src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:
```diff
@@ -3392,5 +3363,85 @@ private void EmitBox(TypeSymbol type, SyntaxNode syntaxNode)
+            LocalDefinition temp = null;
+            if (ptrInvocation.Arguments.Length > 0)
+            {
+                temp = AllocateTemp(ptrInvocation.InvokedExpression.Type, ptrInvocation.Syntax);
```
It's potentially side-effecting. The token for the function pointer must be on top of the stack, but as some of my tests show this can be anything, such as a function call. We need to preserve left-to-right semantics of the code, however, so we need to save it to a temp.

agocke replied:
Cool, can you leave a comment about that? It wasn't obvious how that was side effecting. An example would be helpful

333fred commented on src/Compilers/CSharp/Portable/BoundTree/BoundFunctionPointerInvocation.cs:
```diff
@@ -0,0 +1,22 @@
+        {
+            get
+            {
+                Debug.Assert(InvokedExpression.Type is FunctionPointerTypeSymbol);
```
Because `InvokedExpression.Type` can be null. This is both a null check and a type check.

333fred commented on src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/OverloadResolutionResult.cs:
```diff
@@ -865,12 +869,19 @@ private void ReportDuplicateNamedArgument(MemberResolutionResult<TMember> result
-                (object)delegateTypeBeingInvoked != null ? ErrorCode.ERR_BadDelArgCount :
-                ErrorCode.ERR_BadArgCount;
-            var target = (object)typeContainingConstructor ?? (object)delegateTypeBeingInvoked ?? name;
+            FunctionPointerMethodSymbol functionPointerMethodBeingInvoked = symbols.IsDefault || symbols.Length != 1
```
I could, but I didn't want to enable it for the entire method that I'm not changing.

333fred commented on src/Compilers/CSharp/Test/Symbol/Symbols/FunctionPointerTypeSymbolTests.cs:
```diff
@@ -785,5 +785,76 @@ unsafe void M()
+    void M(delegate*<void> p1,
+           delegate*<string, void> p2,
+           delegate*<string> p3,
+           delegate*<ref string, void> p4)
```
Yes, I'll add a test for them.

333fred commented:
@agocke addressed your feedback.

agocke commented on src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:
```diff
@@ -3392,5 +3363,85 @@ private void EmitBox(TypeSymbol type, SyntaxNode syntaxNode)
+            else if (_ilEmitStyle == ILEmitStyle.Debug)
+            {
+                // The only void methods with usable return values are constructors and we represent those
+                // as BoundObjectCreationExpressions, not BoundCalls.
```
@333fred  How was this one resolved?

333fred replied:
Not sure what happened, I had an assert that I must have accidentally deleted. Fixing.

333fred commented:
@agocke addressed your feedback.

agocke commented on src/Compilers/CSharp/Portable/CodeGen/EmitExpression.cs:
```diff
@@ -3392,5 +3363,90 @@ private void EmitBox(TypeSymbol type, SyntaxNode syntaxNode)
+
+        private void EmitCalli(BoundFunctionPointerInvocation ptrInvocation, UseKind useKind)
+        {
+            EmitExpression(ptrInvocation.InvokedExpression, used: true);
```
lol somehow I completely missed that we were evaluating the receiver expression for the first time here.
